### PR TITLE
nmcli: add idempotent support for any kinds of connections

### LINKED ISSUE #481: nmcli reports changed status even if nothing needs to change
##### SUMMARY

mcli does not know the previous state of an interface so it always reports a status of changed even when nothing does. I would propose parsing the output of "nmcl con show ", then applying the changed and seeing if there are any differences.
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

plugins/modules/net_tools/nmcli.py
##### ANSIBLE VERSION

```paste below
ansible 2.10.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/develop/community.general/ansible-venv/lib64/python3.6/site-packages/ansible
  executable location = /home/user/develop/community.general/ansible-venv/bin/ansible
  python version = 3.6.8 (default, Apr  2 2020, 13:34:55) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```
Previous versions have also been reported as affected.
##### CONFIGURATION

```paste below
# nothing
```

##### OS / ENVIRONMENT

Centos 7

##### STEPS TO REPRODUCE

ansible-playbook nmcli_test.yaml -vvv

`nmcli_test.yaml`:
```yaml
---

- name: nmcli
  hosts: all
  tasks:
  - name: Add IPv4 DNS server addresses
    community.general.nmcli:
      conn_name: System ens192
      type: ethernet
      dns4:
      - 8.8.8.8
      state: present
```

##### EXPECTED RESULTS

changed=0

##### ACTUAL RESULTS

Even though no changes were made, Ansible reports changed=True

```paste below
changed: [192.168.1.70] => {
    "Exists": "Connections do exist so we are modifying them",
    "changed": true,
    "conn_name": "System ens192",
    "invocation": {
        "module_args": {
            "ageingtime": 300,
            "arp_interval": null,
            "arp_ip_target": null,
            "autoconnect": true,
            "conn_name": "System ens192",
            "dhcp_client_id": null,
            "dns4": [
                "8.8.8.8"
            ],
...
        }
    },
    "state": "present"
}
```
Copied from https://github.com/ansible/ansible/issues/29285 but updated to reflect using the community.general module and newer version of Ansible (issue has been reported as far back as 2.0.1).

### PR DESCRIPTION
**Pull request by thinkdoggie**
##### SUMMARY
`nmcli` module does not know the previous state of an interface and always returns a status of "changed" even when nothing does. This PR implemented an idempotent check to modify connections only if changes are detected. The command line arguments and property aliases are slightly different between a variety of connection types.

- Implement show_connection() to retrieve connection profile from command line
- Parse integer enumeration values in show_connection()
- Convert 'bond.options' to alias shortcuts
- Modify connection only if changes are detected
- Support generic alias in during the property comparison

Fixes #481

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins/modules/net_tools/nmcli.py
tests/unit/plugins/modules/net_tools/test_nmcli.py

##### ADDITIONAL INFORMATION
An example of playbook snippet to reproduce this issue:
```yaml
- name: configure network interface
  nmcli:
    conn_name: eth2
    type: ethernet
    mtu: 9000
    ip4: '172.31.250.2/24'
    autoconnect: yes
    state: present
```

### CODE REVIEW

**Comment by thinkdoggie** [comment]
The invoke of "nmcli connection show" in show_connection() breaks the mock assertions in modification related cases. I will revise the unit tests code later.

**Comment by thinkdoggie** [comment]
All previous unit tests have been fixed according to the dependency changes. 

More unit tests are added to `test_nmcli.py` to check idempotent for each type of connections. All newly added features should be covered by these test cases.

**Comment by vlours** [comment]
Hi @thinkdoggie 

Great job on this PR.

My initial code, submitted by @cscheib in #508, included the '--diff' option.
To do so, you will have to store the content of your show_connection function in the result['before'] & result['after'] in a json format. Ansible will compare both result and provide the diff if some values are different.

I've also noted that the function '_compare_conn_params' failed to compare the content of Arrays.
Using multiple values for the DNS:
```
      dns4:
        - 1.1.1.1
        - 8.8.8.8
```
will trigger some changes over and over.

File: plugins/modules/net_tools/nmcli.py [hunk 1]
```diff
@@ -1420,6 +1449,261 @@ def modify_connection(self):
...
+            elif self.ifname is not None:
+                cmd.append(self.ifname)
+            elif self.ip_tunnel_dev is not None:
+                cmd.append('sit%s' % self.ip_tunnel_dev)
+        else:
+            if self.conn_name is not None:
+                cmd.append(self.conn_name)
+            else:
+                raise NmcliModuleError(
+                    "Invalid connection identifier for nmcli")
```
Comment (vlours) [inline plugins/modules/net_tools/nmcli.py:1487]: if self.conn_name is not None:
            cmd.append(self.conn_name)
        elif self.ifname is not None:
            cmd.append(self.ifname)
        else:
            if self.type == 'vlan':
                cmd.append('vlan%s' % to_text(self.vlanid))
            elif self.type == 'vxlan':
                cmd.append('vxlan%s' % to_text(self.vxlan_id))
            elif self.type == 'ipip' or self.type == 'sit':
                cmd.append('ipip%s' % self.ip_tunnel_dev)
            else:
                raise NmcliModuleError(
                    "Invalid connection identifier for nmcli")

**Comment by cspeidel** [comment]
I've closed PR https://github.com/ansible-collections/community.general/pull/508 in favor of this PR, since it is complete for more adapter types.

**Comment by thinkdoggie** [comment]
Hi @cspeidel, thanks for your prior work to inspire me. :-)

**Comment by thinkdoggie** [comment]
Hi @vlours, thanks for your good suggestions. 👍   I'll refine and fix these items later on.

**Comment by cspeidel** [comment]
> Hi @cspeidel, thanks for your prior work to inspire me. :-)

@vlours did all the code for the previous PR, I just migrated the issue and made a PR out of it :-) Thanks to both of you.

File: plugins/modules/net_tools/nmcli.py (hunk 1 above)
Comment (thinkdoggie) [reply plugins/modules/net_tools/nmcli.py]: Code updated

**Comment by thinkdoggie** [comment]
> Hi @thinkdoggie
> 
> Great job on this PR.
> 
> My initial code, submitted by @cscheib in #508, included the '--diff' option.
> To do so, you will have to store the content of your show_connection function in the result['before'] & result['after'] in a json format. Ansible will compare both result and provide the diff if some values are different.
> 
> I've also noted that the function '_compare_conn_params' failed to compare the content of Arrays.
> Using multiple values for the DNS:
> 
> ```
>       dns4:
>         - 1.1.1.1
>         - 8.8.8.8
> ```
> 
> will trigger some changes over and over.

These two work items are resolved in the latest commit.

**Comment by vlours** [comment]
Looks good to me.

Thank you @thinkdoggie for your work and time on this Issue.
 If someone else can review the code, and merge it that will be great.

Cheers,

**Comment by thinkdoggie** [comment]
Hi @felixfontein, could you help review this RP? Thanks!

**Comment by thinkdoggie** [comment]
Hi @opoplawski , could you help to review and verify this PR, as you're the original reporter of this issue at [#29285](https://github.com/ansible/ansible/issues/29285)? :-)

**Comment by thinkdoggie** [comment]
Hi @felixfontein , this PR intends to resolve a problem in the backlog of previous ansible repo [#29285](https://github.com/ansible/ansible/issues/29285) for two years. @vlours, @cspeidel and other guys contributed a lot efforts on both #481 and former [#29285](https://github.com/ansible/ansible/issues/29285). Idempotency is quite essential for each ansible module, no exception for nmcli module. We hope this correction could be included in the next release of community.general.

Could you or someone else help to review and give more advice on this? Thanks!

**Comment by felixfontein** [comment]
@thinkdoggie you need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to). Besides that, I've glanced over the code (I don't know the module's code nor nmcli) and I didn't see something that looks wrong. I'm happy to merge this once it got a changelog fragment.

**Comment by thinkdoggie** [comment]
@felixfontein the changelog fragment is added to this PR. Thanks for your help!

**Comment by felixfontein** [comment]
@thinkdoggie @vlours @cspeidel thanks a lot for working on this and fixing it together! :)

**Comment by opoplawski** [comment]
looks good to  me.  thanks all.

**Pull request merged**